kernel/fs.c

```diff
@@ -286,6 +286,11 @@ int fs_file_write(struct fs_file *file, const char *buffer, uint32_t length, uin
 
 	char *temp = memory_alloc_page(0);
 
+	// if writing past the (current) end of the file, resize the file first
```

Comment:
Careful here: you are changing the size of the file before knowing whether the write succeeds or not.
If some part of the write fails, then the file size needs to be truncated back to the length that succeeded.